sockeye/beam_search.py

```diff
@@ -780,30 +826,53 @@ def forward(self,
 
         t = 1
+        if target_prefix_factors is not None and C.TARGET_FACTOR_SHIFT:
+            target_prefix_factors_shift = pt.zeros(target_prefix_factors.size(0), target_prefix_factors.size(1) + 1, target_prefix_factors.size(2)).to(self.device).type(target_prefix_factors.dtype)
+            target_prefix_factors_shift[:, 1:] = target_prefix_factors
+            target_prefix_factors = target_prefix_factors_shift
+
         for t in range(1, max_iterations + 1):  # max_iterations + 1 required to get correct results
             # (1) obtain next predictions and advance models' state
             # target_dists: (batch_size * beam_size, target_vocab_size)
             # target_factors: (batch_size * beam_size, num_secondary_factors, 2),
             # where last dimension holds indices and scores
-            target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
-                                                                                     model_states,
-                                                                                     vocab_slice_ids)
+            if target_prefix_factors is None:
+                target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
+                                                                                         model_states,
+                                                                                         vocab_slice_ids)
+            else:
+                target_prefix_factor = target_prefix_factors[:, t-1, :] if self.num_target_factors > 1 and t <= target_prefix_factors.size(1) else None
+                target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
+                                                                                         model_states,
+                                                                                         vocab_slice_ids,
+                                                                                         target_prefix_factor)
 
             # (2) Produces the accumulated cost of target words in each row.
             # There is special treatment for finished rows.
             # Finished rows are inf everywhere except column zero, which holds the accumulated model score
             scores, lengths = self._update_scores(target_dists, finished, scores_accumulated,
                                                   lengths, max_output_lengths, pad_dist, eos_dist)
 
+            if target_prefix is not None and t <= target_prefix.size(1):
+                # Make sure search selects the current prefix token by setting the scores of all
+                # other vocabulary items to infinity.
+                target_prefix_in_one_hot = utils.one_hot_encoding_from_prefix(target_prefix[:, t-1:t], self.output_vocab_size)
```

Comment:
I haven't thought about this myself in depth: is there any way we can avoid the one_hot encoding, creating these large vectors?

Reply:
This is actually a very good point. I thought about it and it is easy to reflect my thoughts of how to do this by this example:
```
real target vocab, lets say tensor([ 0,  1,  2,  3,  ...., 15]) (size of (1, 16) 
vocab_slice_ids, lets say tensor([ 0,  1,  2,  3,  6,  9, 10, 12]) (size of (1, 8) 
target prefix id, lets say 9.
```
I guess the key question is to know the index where 9 places in vocab_slice_ids (5 in this example). One we get the index we can simply set it to 1 and the rest is zero. Apparently getting an index can be done by this function https://stackoverflow.com/a/47867513/1953737

I however don't know how fast it can be, especially when we need to do something like: `(t == prefix_id).nonzero(as_tuple=True)` for every prefix id. I used nonzero before and I confirm it is fast (despite nonzero and as_tuple stuff), but it is still not that clear to me that if we do that for all prefix ids in the same timestep, whether it is still that fast ...

Reply:
I have another question about this: what is the current & intended behavior if vocab_slice_ids does not contain ids from the prefix?

If we give preference to the target_prefix, that is, we still force-decode the target prefix id, even though it is not part of `vocab_slice_ids`, wouldn't we need to ids from target prefix to the set of 'allowed' ids in `vocab_slice_ids`?

And then, wouldn't it make more sense to map the target_prefix_ids to the reduced vocabulary space right away before going into the beam search loop? `vocab_slice_ids` is fixed for a single request, so the 'reduced' vocabulary space is the same for each beam search iteration (which is why we cache the reduction operation in the OutputLayer).
Similarly, we could also one-hot-encode all time steps of the target prefix once before the loop, `inf`-mask it, and then only pull out the slice for each time step here. This would reduce memory allocations in each loop and be likely faster.

Reply:
After thinking about this a bit more I agree that there may not be a better way to get prefix ids in the reduced space , other than going through the one-hot.

But I still think the following two changes would be useful:
- ensuring that target prefix ids are part of vocab_slice_ids
- pre-allocating a 'custom scores distribution' (similar like `eos_dist` or `pad_dist` for each target prefix timestep (for cases with and without vocabulary selection). You can do that with `x = pt.full(<shape>, value=np.inf); x[:, target_prefix_ids] = 0` Within the beam search loop, your logic would then simplify to something like `scores += x`.
- This would also allow you to put this logic into `UpdateScores`.

What do you think?

In regards to finding the index of a prefix id in the reduced space:
- ids in `vocab_slice_ids` are unique and sorted by increasing id (plus padded to a multiple of 8) 
- if we assume target prefix ids are a strict subset of `vocab_slice_ids` (we first need to make sure of that), isnt the 'reduced id' of a target prefix id simply the position in vocab_slice

Reply:
These discussions are very useful. Thanks Felix.

1. "what is the current & intended behavior if vocab_slice_ids does not contain ids from the prefix?" -> At the moment I just assume what in the prefix will be in the vocab_slice_ids. But you are right this is not great.
 
2. "ensuring that target prefix ids are part of vocab_slice_ids" -> Done I added that now.

3. "pre-allocating a 'custom scores distribution' (similar like eos_dist or pad_dist for each target prefix timestep (for cases with and without vocabulary selection). You can do that with x = pt.full(<shape>, value=np.inf); x[:, target_prefix_ids] = 0 Within the beam search loop, your logic would then simplify to something like scores += x. This would also allow you to put this logic into UpdateScores." 

I got your idea here and I need to think more about it. It is true that sum with np.inf is used with `pad_dist` and `pt.where()`. It was actually a bit difficult for me to understand these logics in the beginning but when I got it I noticed it is very smart. Meanwhile, my thought at the moment is also that I don't really know the benefit of adding sum instead of doing in-place `masked_fill_`. Doing masked_fill is intuitive, and it is an in-place operation so it is fast and I notice people do very often in Pytorch. Another thought is how to cover all cases of arbitrary length of prefix with this. Finally with the approach of using the one hot encoding method at the moment I found that we are able handle both prefix and prefix factors smoothly. I am not sure how clean it is to apply sum of np.inf for both constraining prefix and prefix factors.

Meanwhile I do notice from this discussion is that doing `one_hot_encoding_from_prefix` for each time step separately is inefficient (I did that for every time-step because I thought doing it once for the full tensor results in a tensor of size of batch_size * max_length* vocab size, which could be a big issue to memory). But since max_len of prefix is often small (I believe not a lot of people use more than, say, 5 tokens to say the least), it should be OK regarding memory usage. So I now call the function for both a full size of prefix and prefix factors. Please take a look and let me know if this is cleaner from your point of view.

Reply:
After thinking more about this, I think I agreed with you about several things:

1. `And then, wouldn't it make more sense to map the target_prefix_ids to the reduced vocabulary space right away before going into the beam search loop? vocab_slice_ids is fixed for a single request, so the 'reduced' vocabulary space is the same for each beam search iteration (which is why we cache the reduction operation in the OutputLayer).
Similarly, we could also one-hot-encode all time steps of the target prefix once before the loop, inf-mask it, and then only pull out the slice for each time step here. This would reduce memory allocations in each loop and be likely faster.` -> It is true that we should map the target_prefix_ids to the reduced vocabulary space right away before going into the beam search loop. This is a very obvious thing I should do and somehow I missed it indeed.

2. `You can do that with x = pt.full(<shape>, value=np.inf); x[:, target_prefix_ids] = 0 Within the beam search loop, your logic would then simplify to something like scores += x.` I agreed that while mask_filled_ works perfectly fine, it is better here to change it as simple as `scores += x`. It is better because of at least two reasons (1): it looks simpler indeed. and (2): it is consistent because it follows our practice of similar stuff (+ first_step_mask, + scores in _update_scores). Furthermore, the logic here can be applied exactly to prefix factors masking. So yeah I like this idea much more now. So I changed our code based on your suggestion.

BTW Regarding (2): I also changed the generate_one_hot_function to `gen_prefix_masking` instead, as we now create a new matrix which are inf everywhere except zero for prefix ids. But the logic is still the same for that function (i.e. gen_prefix_masking) because it is based on the pt.nn.functional.one_hot. So everything is ``smooth" here.

Meanwhile I still prefer keeping +x separately instead of putting it into UpdateScores at the moment. There are two reasons here. 1. In greedy module we don't have `updatescores` stuff. So keeping the same line of code (+x) between two modules makes things consistency. 2. tracing `UpdateScores` seems does not gain if not hurt on GPUs so at least there is no need yet to add stuff into UpdateScores at the moment for a better performance.